src/components/input_fields/code_field.tsx

```diff
@@ -0,0 +1,296 @@
+     * Checks whether the given character value is allowed on this field.
+     * @param charValue - The value of the character.
+     * @private
+     */
+    private _isKeyAllowed(charValue: string) {
+        const { type } = this.props;
+        const regex = {
+            text: /\w/,
+            number: /\d/,
+            password: /[ -~]/,  // match any ascii printable characters
+        }
+        return !charValue || regex[type].test(charValue);
+    }
+
+    /**
+     * Actually changes the value of of the code inputs.
+     * @param index - The index of the field.
+     * @param charValue - The value of the character typed. Empty string is fine. Null will completely remove the value.
+     * @param focusOnIndex - The next input field to be selected. -1 to default it to {index}.
+     * @return - Whether the change was successful.
+     */
+    private _changeValue(index: number, charValue: string, focusOnIndex: number): boolean {
+        const { onChangeHandler, id } = this.props;
+        const { values } = this.state;
+
+        if (values[index] === charValue) {
+            return true;
+        } else if (!this._isKeyAllowed(charValue)) {
+            return false;
+        }
+
+        if (charValue === null) {
+            // charValue null, remove the character completely, add one to the end to make sure the length is OK.
+            values.splice(index, 1);
+            values.push('');
+        } else {
+            values[index] = charValue;
+        }
+
+        // Default chars to _emptyChar, remove the empty spaces in the end
+        const newValue: string = values.map(v => v || CodeField._emptyChar).join("").replace(/\s+$/g, "");
+        onChangeHandler(id, newValue);
+
+        this.setState({ touched: true, values });
+        this._focusOnPromptInput(focusOnIndex >= 0 ? focusOnIndex : index);
+        return true;
+    }
+
+    /**
+     * Gets a function that will handle special keyboard keys.
+     * @param index - The index of the field in which the key was pressed.
+     * @param key - The name of the keyboard key that was pressed.
+     * @return - Null if the key was special; otherwise a function that will return true if the default behavior should
+     * be blocked;
+     */
+    private _getSpecialKeyHandler(index: number, key: string): () => boolean | null {
+        const { type, maxLength } = this.props;
+        const { values } = this.state;
+        /** a mapping of keys that should be handled. Each function returns whether the event was handled. **/
+        return {
+            // arrow left & right let you move between the fields
+            ArrowLeft: () => this._focusOnPromptInput(index - 1),
+            ArrowRight: () => this._focusOnPromptInput(index + 1),
+            // arrow up / down let you increase / decrease numbers
+            ArrowUp: () => type == 'number' &&
+                this._changeValue(index, values[index] && String.fromCharCode(values[index].charCodeAt(0) + 1), -1),
```

Comment:
There is a bug here once the number goes above 9 or below 1 it starts to act weird